_drafts/cortexm-and-clang.md

```diff
@@ -0,0 +1,971 @@
+ + core.DivideZero               Check for division by zero
+ + core.DynamicTypePropagation   Generate dynamic type information
+ + core.NonNullParamChecker      Check for null pointers passed as arguments to a function whose arguments are references or marked with the 'nonnull' attribute
+ + core.NullDereference          Check for dereferences of null pointers
+ + core.StackAddressEscape       Check that addresses to stack memory do not escape the function
+ + core.UndefinedBinaryOperatorResult Check for undefined results of binary operators
+[...]
+NOTE: "+" indicates that an analysis is enabled by default.
+```
+
+I find the default set to be a good starting point but you can easily enable non-default ones or
+disable individual defaults. (You can even implement your own custom pugins but these topics are outside the scope of this article.)
+
+Let's run scan-build on our example project. We'll see it finds a few interesting potential issues that neither Clang or GCC noticed during compilation.
+
+Results will be emitted directly to the console or you can browse them using the `scan-view`
+command bundled with LLVM.
+
+#### Running scan-build
+
+```bash
+$ cd interrupt/examples/freertos-example-llvm/
+$ make clean
+$ scan-build make
+scan-build: Using 'bin/clang-10' for static analysis
+[...]
+Linking library
+Generated build/nrf52.elf
+scan-build: 8 bugs found.
+scan-build: Run 'scan-view /var/folders/dm/b0yt' to examine bug reports.
+```
+
+#### core.NullDereference Checker
+
+```c
+int example_operate_on_pointer(uint8_t *pointer) {
+  int result = 0;
+  if (pointer == NULL) {
+    result = -1;
+  }
+  result += *pointer;
+  return result;
+}
+```
+
+```bash
+src/builtin_scanbuild_examples.c:13:13:
+  warning: Dereference of null pointer (loaded from variable 'pointer')
+  result += *pointer;
+```
+
+> NOTE: The NullDereference Checker only looks for null pointer accesses _after_ there has already been a nullability check. This helps prevent the generation of false positives for functions that never get passed a NULL pointer.
+
+#### core.DivideZero Checker
+
+```c
+int example_divide_by_zero(int denominator) {
+  int rv = 5;
+  if (denominator == 0) {
+    rv = 1 / denominator;
+  }
+  return rv;
+}
+```
+
+```bash
+src/builtin_scanbuild_examples.c:20:11: warning: Division by zero
+  rv = 1 / denominator;
+```
+
+### Enabling "Malloc" Static Analysis Checkers
+
+`scan-build` also includes a sophisticated suite of dynamic memory checkers to catch bugs such as
+memory leaks, use-after-free and double free (These are mostly run as part of the `unix.Malloc` checker). If you use the C standard library calls (i.e
+`calloc`, `malloc`, `free`, etc) in your embedded project, these checks will work out of the box.
+
+However, for many embedded projects this is not the case. It's typical to have a heap (or "byte
```

Comment:
However, many embedded projects might need to instrument this checker to work with the project codebase. 

Or something like that.